pilot/pkg/bootstrap/istio_ca.go

```diff
@@ -263,6 +266,122 @@ func (s *Server) loadRemoteCACerts(caOpts *caOptions, dir string) error {
 	return nil
 }
 
+// isValidCACertsFile As we are watching entire directory, but interested
+// in only 'ca-key.pem', 'ca-cert.pem', 'root-cert.pem' and 'cert-chain.pem'.
+// Events on other files are ignored.
+func isValidCACertsFile(path string) bool {
+	_, file := filepath.Split(path)
+
+	for _, name := range []string{"ca-cert.pem", "ca-key.pem", "root-cert.pem", "cert-chain.pem"} {
+		if file == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+// handleEvent handles the events on cacerts related files.
+// If create/write(modified) event occurs, then it verifies that
+// newly introduced cacerts are intermediate CA which is generated
+// from cuurent root-cert.pem. Then it updates and keycertbundle
+// and generates new dns certs.
+// TODO(rveerama1): Add support for new ROOT-CA rotation also.
+func handleEvent(s *Server) {
+	log.Info("Update Istiod cacerts")
+
+	currentCABundle := s.CA.GetCAKeyCertBundle().GetRootCertPem()
+	newCABundle, err := ioutil.ReadFile(path.Join(LocalCertDir.Get(), "root-cert.pem"))
+	if err != nil {
+		log.Error("failed reading root-cert.pem: ", err)
+		return
+	}
+
+	// Only updating intermediate CA is supported now
+	if !bytes.Equal(currentCABundle, newCABundle) {
+		log.Info("Updating new ROOT-CA not supported")
+		return
+	}
+
+	err = s.CA.GetCAKeyCertBundle().UpdateVerifiedKeyCertBundleFromFile(
+		path.Join(LocalCertDir.Get(), "ca-cert.pem"),
+		path.Join(LocalCertDir.Get(), "ca-key.pem"),
+		path.Join(LocalCertDir.Get(), "cert-chain.pem"),
+		path.Join(LocalCertDir.Get(), "root-cert.pem"))
+	if err != nil {
+		log.Error("Failed to update new Plug-in CA certs: ", err)
+		return
+	}
+
+	err = s.updatePluggedinRootCertAndGenKeyCert()
+	if err != nil {
+		log.Error("Failed generating plugged-in istiod key cert: ", err)
+		return
+	}
+}
+
+// handleCACertsFileWatch handles the events on cacerts files
+func (s *Server) handleCACertsFileWatch() {
+	var timerC <-chan time.Time
+	for {
+		select {
+		case <-timerC:
+			timerC = nil
+			handleEvent(s)
+
+		case event, ok := <-s.cacertsWatcher.Events:
+			if !ok {
+				log.Info("Failed to catch events on cacerts files")
+				continue
+			}
+
+			if event.Op&fsnotify.Write == fsnotify.Write || event.Op&fsnotify.Create == fsnotify.Create {
+				valid := isValidCACertsFile(event.Name)
+				if valid && timerC == nil {
+					timerC = time.After(100 * time.Millisecond)
+				}
+			}
+
+		case err := <-s.cacertsWatcher.Errors:
+			if err != nil {
+				log.Error("Failed to catch events on cacerts file: ", err)
+			}
+		}
+	}
+}
+
+func (s *Server) addCACertsFileWatcher(dir string) error {
+	err := s.cacertsWatcher.Add(dir)
+	if err != nil {
+		log.Info("Failed to add file watcher: ", err)
```

Comment:
change to error and add a message that PILOT_AUTO_RELOAD_PLUGIN_CERTS will not work.

Reply:
Done